### PR TITLE
Add Linux VM extension bootstrap script and conditions

### LINKED ISSUE #603: Bootstrap failure detection
/kind feature

**Describe the solution you'd like**
If bootstrapping doesn't succeed, set AzureMachine.Status.FailureReason and FailureMessage to indicate there was an error. 

**Anything else you would like to add:**
More information in https://github.com/kubernetes-sigs/cluster-api/issues/2554.

We may need to amend the bootstrap provider contract to require them to write a sentinel file indicating success to a specific location because not all bootstrap providers will necessarily use cloud-init and have that as a consistent means for checking success/failure. I will probably write up a separate proposal for that.

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**

**What this PR does / why we need it**: Implements bootstrap failure detection using conditions and VM extensions as proposed in #1076.

This is what a user will see when an AzureMachine has been successfully provisioned but has not yet finished running the bootstrap script:
```
$ k get azuremachine default-template-md-0-w78jt
NAME                                       READY   STATE
default-template-md-0-w78jt                false   Updating
```

```
status:
    conditions:
  - lastTransitionTime: "2021-03-17T00:02:14Z"
    reason: VMUpdating
    severity: Info
    status: "False"
    type: Ready
  - lastTransitionTime: "2021-03-17T00:02:14Z"
    reason: BootstrapInProgress
    severity: Info
    status: "False"
    type: BoostrapSucceeded
  - lastTransitionTime: "2021-03-17T00:02:14Z"
    reason: VMUpdating
    severity: Info
    status: "False"
    type: VMRunning
```

After the bootstrap script has executed successfully, the AzureMachine status shows:

```
$ k get azuremachine default-template-md-0-w78jt
NAME                                       READY   STATE
default-template-md-0-w78jt                true    Succeeded
```

```
status:
  conditions:
  - lastTransitionTime: "2021-03-16T23:56:17Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2021-03-16T23:56:17Z"
    status: "True"
    type: BoostrapSucceeded
  - lastTransitionTime: "2021-03-16T23:56:17Z"
    status: "True"
    type: VMRunning
```

If for some reason the bootstrap script fails to execute, the status will show:

```
$ k get azuremachines default-template-md-0-ppjmh
NAME                                       READY   STATE
default-template-md-0-ppjmh                false   Failed
```

```
  conditions:
  - lastTransitionTime: "2021-03-17T00:22:39Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2021-03-17T00:22:39Z"
    reason: BootstrapFailed
    severity: Error
    status: "False"
    type: BoostrapSucceeded
  - lastTransitionTime: "2021-03-17T00:22:39Z"
    status: "True"
    type: VMRunning
```

Also renames type "VMState" to "ProvisioningState" to be more generic and match Azure naming. **This is a breaking change** for anyone importing CAPZ types, but should have no impact on user templates and existing CRDs (conversion from v1alpha3 to v1alpha4 is auto-generated).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #603 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Add Linux VM extension bootstrap script and conditions
Renames type "VMState" to "ProvisioningState"
```

### CODE REVIEW

File: azure/scope/machine.go [hunk 1]
```diff
@@ -382,6 +385,40 @@ func (m *MachineScope) SetFailureReason(v capierrors.MachineStatusError) {
...
+	if value {
+		conditions.MarkTrue(m.AzureMachine, condition)
+	} else {
+		conditions.MarkFalse(m.AzureMachine, condition, reason, severity, "")
+	}
+}
+
+func (m *MachineScope) UpdateStatus() {
+	switch m.VMState() {
+	case infrav1.VMStateSucceeded:
```
Comment (shysank) [inline azure/scope/machine.go:399]: I like the idea of movie this logic to scope. Just a side note not related to the pr, as we add more stuff to scope, we may want to prioritize #1108

File: azure/services/vmextensions/client.go [hunk 2]
```diff
@@ -61,19 +61,11 @@ func (ac *azureClient) Get(ctx context.Context, resourceGroupName, vmName, name
 }
 
 // CreateOrUpdate creates or updates the virtual machine extension
```
Comment (shysank) [inline azure/services/vmextensions/client.go:63]:
```suggestion
// CreateOrUpdateAsync creates or updates the virtual machine extension.
```

File: azure/services/vmextensions/vmextensions.go [hunk 3]
```diff
@@ -55,13 +58,29 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	defer span.End()
 
 	for _, extensionSpec := range s.Scope.VMExtensionSpecs() {
-		if _, err := s.client.Get(ctx, s.Scope.ResourceGroup(), extensionSpec.VMName, extensionSpec.Name); err == nil {
-			// check for the extension and don't update if already exists
-			// TODO: set conditions based on extension status
+		if existing, err := s.client.Get(ctx, s.Scope.ResourceGroup(), extensionSpec.VMName, extensionSpec.Name); err == nil {
+			// check the extension status and set the associated conditions.
+			switch compute.ProvisioningState(to.String(existing.ProvisioningState)) {
```
Comment (shysank) [inline azure/services/vmextensions/vmextensions.go:63]: Can we follow a similar pattern like we do when updating VM state and move this logic to scope?, ie.
```
scope.setVMExtState()
scope.updateStatus()
```

Maybe we can do it as part of future refactoring.

File: azure/services/vmssextensions/vmssextensions.go [hunk 4]
```diff
@@ -55,32 +57,26 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	defer span.End()
 
 	for _, extensionSpec := range s.Scope.VMSSExtensionSpecs() {
-		if _, err := s.client.Get(ctx, s.Scope.ResourceGroup(), extensionSpec.ScaleSetName, extensionSpec.Name); err == nil {
-			// check for the extension and don't update if already exists
-			// TODO: set conditions based on extension status
-			continue
+		if existing, err := s.client.Get(ctx, s.Scope.ResourceGroup(), extensionSpec.ScaleSetName, extensionSpec.Name); err == nil {
+			// check the extension status and set the associated conditions.
+			switch compute.ProvisioningState(to.String(existing.ProvisioningState)) {
```
Comment (shysank) [inline azure/services/vmssextensions/vmssextensions.go:62]: same comment as vmextension.

File: exp/controllers/azuremachinepool_controller.go [hunk 5]
```diff
@@ -297,12 +298,15 @@ func (r *AzureMachinePoolReconciler) reconcileNormal(ctx context.Context, machin
 	switch machinePoolScope.ProvisioningState() {
```
Comment (shysank) [inline exp/controllers/azuremachinepool_controller.go:298]: not part of this pr, but we should probably move this to machinepoolscope.

File: azure/defaults.go [hunk 6]
```diff
@@ -65,6 +65,11 @@ const (
 	ControlPlaneNodeGroup = "control-plane"
 )
 
+const (
+	// BootstrapExtensionCommand is the command that runs on the Boostrap VM extension to check for bootstrap success
+	BootstrapExtensionCommand = "for i in $(seq 1 240); do test -f /run/cluster-api/bootstrap-success.complete && break; if [ $i -eq 240 ]; then return 1; else sleep 5; fi; done"
```
Comment (shysank) [inline azure/defaults.go:70]: can we assign 240 to a constant for more clarity?
maybe more notes on `/run/cluster-api/bootstrap-success.complete` - where this comes from and what it does?

File: azure/services/vmssextensions/vmssextensions.go [hunk 7]
```diff
@@ -55,32 +57,26 @@ func (s *Service) Reconcile(ctx context.Context) error {
...
+				s.Scope.V(4).Info("extension provisioning state is failed", "vm extension", extensionSpec.Name, "scale set", extensionSpec.ScaleSetName)
+				s.Scope.SetCondition(infrav1.BootstrapSucceededCondition, infrav1.BootstrapFailedReason, clusterv1.ConditionSeverityError, false)
+				return errors.New("extension state failed")
+			}
+		} else if !azure.ResourceNotFound(err) {
+			return errors.Wrapf(err, "failed to get vm extension %s on scale set %s", extensionSpec.Name, extensionSpec.ScaleSetName)
 		}
 
-		s.Scope.V(2).Info("creating VMSS extension", "vssm extension", extensionSpec.Name)
-		err := s.client.CreateOrUpdate(
```
Comment (shysank) [inline azure/services/vmssextensions/vmssextensions.go:65]: should we also remove the client code for this?

**Comment by shysank** [comment]
Just realized it's a wip pr, please feel free to ignore the review comments that are not relevant.

File: exp/controllers/azuremachinepool_controller.go (hunk 5 above)
Comment (CecileRobertMichon) [reply exp/controllers/azuremachinepool_controller.go:298]: #819 does a lot of rework of this area so didn't want to conflict too much, but yes agreed

File: azure/services/vmextensions/client.go [hunk 8]
```diff
@@ -60,20 +60,12 @@ func (ac *azureClient) Get(ctx context.Context, resourceGroupName, vmName, name
...
-	future, err := ac.vmextensions.CreateOrUpdate(ctx, resourceGroupName, vmName, name, parameters)
-	if err != nil {
-		return err
-	}
-	err = future.WaitForCompletionRef(ctx, ac.vmextensions.Client)
-	if err != nil {
-		return err
-	}
-	_, err = future.Result(ac.vmextensions)
+	_, err := ac.vmextensions.CreateOrUpdate(ctx, resourceGroupName, vmName, name, parameters)
```
Comment (CecileRobertMichon) [inline azure/services/vmextensions/client.go:68]: @devigned thoughts on this approach of doing "async" reconciliation? Is it significantly less efficient than storing the future or since we only care about eventually getting the extension status do you think it's appropriate in this situation?

File: azure/scope/machinepool.go [hunk 9]
```diff
@@ -259,6 +262,24 @@ func (m *MachinePoolScope) SetFailureReason(v capierrors.MachineStatusError) {
 	m.AzureMachinePool.Status.FailureReason = &v
 }
 
+func (m *MachinePoolScope) SetBootstrapConditions(provisioningState string, extensionName string) error {
+	switch infrav1.ProvisioningState(provisioningState) {
+	case infrav1.Succeeded:
+		m.V(4).Info("extension provisioning state is succeeded", "vm extension", extensionName, "scale set", m.Name())
+		conditions.MarkTrue(m.AzureMachinePool, infrav1.BootstrapSucceededCondition)
```
Comment (CecileRobertMichon) [inline azure/scope/machinepool.go:269]: it'd be nice to set conditions per VMSS VM, hopefully that gets easier with #819

**Comment by CecileRobertMichon** [comment]
Thanks for the initial review @shysank! I've addressed most of your comments. I still have docs and unit tests as TODOs.

File: azure/scope/machinepool.go (hunk 9 above)
Comment (CecileRobertMichon) [reply azure/scope/machinepool.go:269]: Right now it's at the AzureMachinePool.Status level:
```
status:
    conditions:
    - lastTransitionTime: "2021-03-16T23:53:39Z"
      status: "True"
      type: BoostrapSucceeded
    - lastTransitionTime: "2021-03-16T23:49:37Z"
      status: "True"
      type: ScaleSetRunning
```

File: api/v1alpha4/types.go [hunk 10]
```diff
@@ -203,24 +203,28 @@ type PublicIPSpec struct {
...
+	// Failed ...
+	Failed ProvisioningState = "Failed"
+	// Migrating ...
+	Migrating ProvisioningState = "Migrating"
+	// Succeeded ...
+	Succeeded ProvisioningState = "Succeeded"
+	// Updating ...
+	Updating ProvisioningState = "Updating"
+	// Deleted represents a deleted VM
 	// NOTE: This state is specific to capz, and does not have corresponding mapping in Azure API (https://docs.microsoft.com/en-us/azure/virtual-machines/states-lifecycle#provisioning-states)
```
Comment (alexeldeib) [inline api/v1alpha4/types.go:226]: dead link, replace with https://docs.microsoft.com/en-us/azure/virtual-machines/states-billing#provisioning-states

File: azure/services/vmextensions/client.go (hunk 8 above)
Comment (CecileRobertMichon) [reply azure/services/vmextensions/client.go:68]: I'm thinking in the future we can move this over to VM CreateOrUpdate once that's async. For now the extension could be a long-running operation (timeout is 20 minutes) so we don't want that to be part of a blocking operation

**Comment by CecileRobertMichon** [comment]
This is ready for review now

**Comment by nader-ziada** [comment]
when I try running it, I don't see the `Updating` phase

```
NAME                           READY   STATE
nz-test1-control-plane-wk926   true    Succeeded
nz-test1-md-0-gc5gc
```

then goes directly to 

```
NAME                           READY   STATE
nz-test1-control-plane-wk926   true    Succeeded
nz-test1-md-0-gc5gc            true    Succeeded
```

not sure why yet, will dig into it a bit more

File: azure/services/vmextensions/client.go (hunk 8 above)
Comment (devigned) [reply azure/services/vmextensions/client.go:68]: I think this should be fine. Though, it would be easy to call this multiple times if one is not tracking an ongoing operation via a secret. With the AMP future, we check to see if the operation is done prior to proceeding, which the Azure SDK for Go can help with. In this case, you'd be responsible for managing that state.

**Review by devigned — COMMENTED** [review]
Looks great. Only had a minor comment about async APIs.

**Comment by CecileRobertMichon** [comment]
> when I try running it, I don't see the Updating phase

@nader-ziada I wonder if you're doing something differently, this is what I see when scaling up:

```
Every 2.0s: kubectl get azuremachines                                                                                                 
Ceciles-MacBook-Pro.local: Tue Mar 23 18:42:51 2021

NAME                                   READY   STATE
default-template-control-plane-5zq49   true    Succeeded
default-template-md-0-b9cpb            true    Succeeded
default-template-md-0-bnq44
default-template-md-0-vkd4b            true    Succeeded
```

```
Every 2.0s: kubectl get azuremachines                                                                                                 
Ceciles-MacBook-Pro.local: Tue Mar 23 18:43:08 2021

NAME                                   READY   STATE
default-template-control-plane-5zq49   true    Succeeded
default-template-md-0-b9cpb            true    Succeeded
default-template-md-0-bnq44            false   Updating
default-template-md-0-vkd4b            true    Succeeded
```

```
Every 2.0s: kubectl get azuremachines                                                                                                 
Ceciles-MacBook-Pro.local: Tue Mar 23 18:43:38 2021

NAME                                   READY   STATE
default-template-control-plane-5zq49   true    Succeeded
default-template-md-0-b9cpb            true    Succeeded
default-template-md-0-bnq44            true    Succeeded
default-template-md-0-vkd4b            true    Succeeded
```

That first step I'm hoping we can change to `default-template-md-0-bnq44            false   Creating` as part of #1067 (by returning and setting status before the VM is done creating)

**Comment by nader-ziada** [comment]
I tried it again and seems to work as expected
```
NAME                                READY   STATE
nz-test1-control-plane-pq8lt        true    Succeeded
nz-test1-md-0-qs2nf                 false   Updating
nz-test1-md-0-qs2nf                 false   Updating
nz-test1-md-0-qs2nf                 true    Succeeded
```

**Comment by nader-ziada** [comment]
/lgtm

@devigned any comments?

**Review by devigned — APPROVED** [review]
/lgtm
/approve
